neurodamus/utils/memory.py

```diff
@@ -160,6 +161,159 @@ def pretty_printing_memory_mb(memory_mb):
+    """
+    Distributes cells across ranks based on their memory load.
+
+    This function uses a greedy algorithm to distribute cells across ranks such that
+    the total memory load is balanced. Cells with higher memory load are distributed first.
+
+    Args:
+        dry_run_stats (DryRunStats): A DryRunStats object.
+        num_ranks (int): The number of ranks.
+
+    Returns:
+        rank_allocation (dict): A dictionary where keys are rank IDs and
+                                values are lists of cell IDs assigned to each rank.
+        rank_memory (dict): A dictionary where keys are rank IDs
+                            and values are the total memory load on each rank.
+    """
+    logging.debug("Distributing cells across %d ranks", num_ranks)
+
+    def validate_inputs(dry_run_stats, num_ranks, batch_size):
+        assert isinstance(dry_run_stats, DryRunStats), "dry_run_stats must be a DryRunStats object"
+        assert isinstance(num_ranks, int), "num_ranks must be an integer"
+        assert num_ranks > 0, "num_ranks must be a positive integer"
+        assert isinstance(batch_size, int), "batch_size must be an integer"
+        assert batch_size > 0, "batch_size must be a positive integer"
+        set_metype_gids = set()
+        for values in dry_run_stats.metype_gids.values():
+            set_metype_gids.update(values.keys())
+        assert set_metype_gids == set(dry_run_stats.metype_memory.keys())
+        average_syns_keys = set(dry_run_stats.average_syns_per_cell.keys())
+        metype_memory_keys = set(dry_run_stats.metype_memory.keys())
+        assert average_syns_keys == metype_memory_keys
+
+    # Check inputs
+    validate_inputs(dry_run_stats, num_ranks, batch_size)
+
+    # Multiply the average number of synapses per cell by 2.0
+    # This is done since the biggest memory load for a synapse is 2.0 kB and at this point in the
+    # code we have lost the information on whether they are excitatory or inhibitory
+    # so we just take the biggest value to be safe. (the difference between the two is minimal)
+    average_syns_mem_per_cell = {k: v * 2.0 for k, v in dry_run_stats.average_syns_per_cell.items()}
+
+    # Prepare a list of tuples (cell_id, memory_load)
+    # We sum the memory load of the cell type and the average number of synapses per cell
+    def generate_cells(metype_gids):
+        for cell_type, gids in metype_gids.items():
+            for gid in gids:
+                memory_usage = (dry_run_stats.metype_memory[cell_type] +
+                                average_syns_mem_per_cell[cell_type])
+                yield gid, memory_usage
+
+    # Initialize structures
+    ranks = [(0, i) for i in range(num_ranks)]  # (total_memory, rank_id)
+    heapq.heapify(ranks)
+    rank_allocation = {}
+    rank_memory = {}
+
+    def assign_cells_to_rank(rank_id, batch_memory, rank_allocation, total_memory,
+                             rank_memory, pop):
```

Comment:
```suggestion
    def assign_cells_to_rank(rank_id, batch_memory, total_memory, pop):
```
The other variables would be in scope for this function. Simplifies call structure / code a bit.

Reply:
Will integrate into next push, thanks!